electrum/lnworker.py

```diff
@@ -1275,6 +1278,24 @@ def save_payment_info(self, info: PaymentInfo) -> None:
             self.payments[key] = info.amount, info.direction, info.status
         self.wallet.save_db()
 
+    def add_received_htlc(self, htlc, expected_msat):
+        status = self.get_payment_status(htlc.payment_hash)
+        if status == PR_PAID:
+            return True, None
+        s = self.mpp_htlcs[htlc.payment_hash]
+        if htlc not in s:
+            s.add(htlc)
+        total = sum([htlc.amount_msat for htlc in s])
+        first_timestamp = min([htlc.timestamp for htlc in s])
+        expired = time.time() - first_timestamp > MPP_EXPIRY
```

Comment:
If I understand correctly, only a new incoming HTLC would trigger the expiry of the total payment, but we would hold on to the already accepted HTLCs indefinitely if the last part of the payment never arrives. A possible solution to this could be adding a coroutine that watches for all the incoming parts and cancels once a time limit is reached.

Reply:
indeed, good point.